homeassistant/components/husqvarna_automower/calendar.py

```diff
@@ -0,0 +1,95 @@
+"""Creates a calendar entity for the mower."""
+
+from datetime import datetime
+import logging
+from zoneinfo import ZoneInfo
+
+from homeassistant.components.calendar import CalendarEntity, CalendarEvent
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util import dt as dt_util
+
+from . import AutomowerConfigEntry
+from .coordinator import AutomowerDataUpdateCoordinator
+from .entity import AutomowerBaseEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: AutomowerConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up lawn mower platform."""
+    coordinator = entry.runtime_data
+    async_add_entities(
+        AutomowerCalendarEntity(mower_id, coordinator) for mower_id in coordinator.data
+    )
+
+
+class AutomowerCalendarEntity(AutomowerBaseEntity, CalendarEntity):
+    """Representation of the Automower Calendar element."""
+
+    _attr_name: str | None = None
+
+    def __init__(
+        self,
+        mower_id: str,
+        coordinator: AutomowerDataUpdateCoordinator,
+    ) -> None:
+        """Set up AutomowerCalendarEntity."""
+        super().__init__(mower_id, coordinator)
+        self._attr_unique_id = mower_id
+        self._event: CalendarEvent | None = None
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the current or next upcoming event."""
+        schedule = self.mower_attributes.calendar
+        if schedule.timeline is None:
+            return None
+        cursor = schedule.timeline.active_after(dt_util.now())
+        program_event = next(cursor, None)
+        _LOGGER.debug("program_event %s", program_event)
+        if not program_event:
+            return None
+        return CalendarEvent(
+            summary=program_event.program_id,
+            start=program_event.start.replace(
```

Comment:
I didn't review the entire aiomower library, but one thing to watch out for: The python / linux system timezone may be different that the home assistant timezone. I found in ical i needed to pass in `dt_util.DEFAULT_TIME_ZONE` or `start_date.tzinfo` (for `async_get_events`) and consistently use that timezone for all `datetime.now` calls to avoid bugs where things get unexpected moved around between utc and the home assistant tz.  It is fairly common to see home assistant users who have a docker container set to `UTC` which python will use and some other timezone for home assistant.

I don't see a specific bug necessarily, and didn't look for one, but generally my thought here is: Why is it safe to completely ignore the timezone of this `datetime` object and use the time if it were a naive timestamp without any timezone. Perhaps this works fine if it is consistent everywhere, but in that case it might be simpler to pass in the timezone to the client library up front to use.

I'm not necessarily against this code as is, just wanted to raise it as a discussion point.

Reply:
Yes, I'm aware of the problems with the different timezones. I've experienced this by myself with the devcontainer(UTC) and the production environment (local timezone).

> I don't see a specific bug necessarily, and didn't look for one, but generally my thought here is: Why is it safe to completely ignore the timezone of this datetime object and use the time if it were a naive timestamp without any timezone

It is a naive datetime object, because the mower is not aware of it's timezone and also API and the library not. Do you think we should pass the timezone when initializing the session, to the library?
I'm already working on that, but im getting stucked, when passing the timezone info from the session to the mashumaro serialization.
https://github.com/Thomas55555/aioautomower/pull/381/files#diff-9749274a6e273ee74d61048a94b1912aafc44e915492cc23bf528f7e7c7401af

Reply:
OK If these are all naive timezones, then that helps and there won't be conversion bugs. I think that is ok.